include/alpaka/core/ApiHipRt.hpp

```diff
@@ -182,8 +182,21 @@ namespace alpaka
         static inline Error_t freeAsync([[maybe_unused]] void* devPtr, [[maybe_unused]] Stream_t stream)
         {
             // hipFreeAsync is implemented only in ROCm 5.2.0 and later.
-#    if HIP_VERSION >= 50'200'000
+#    if HIP_VERSION >= 50'600'000
             return ::hipFreeAsync(devPtr, stream);
+#    elif HIP_VERSION >= 50'300'000
+            // before ROCm 5.6.0, hipFreeAsync fails on a null pointer deallocation
+            if(devPtr)
+            {
+                return ::hipFreeAsync(devPtr, stream);
+            }
+            else
+            {
+                return ::hipSuccess;
+            }
+#    elif HIP_VERSION >= 50'200'000
+            // in ROCm 5.2.x, hipFreeAsync makes a subsequent hipStreamDestroy hang
+            return ::hipFree(devPtr);
```

Comment:
IMO this is not correct. You should wait until all work in the stream is finished and then you can call hipFree, else a kernel that is using the pointer can still be active.
Another way is to enqueue a host task into the stream which is executing `hipFree`. Then it will be asynchronous too.
The second option which IMO is nicer because it keeps the asynchronous execution of this function requires that the workaround is moved to the place where `api::freeAsync` is called.

Reply:
> IMO this is not correct. You should wait until all work in the stream is finished and then you can call hipFree, else a kernel that is using the pointer can still be active.

Actually, `hipFree` itself is blocking, so the behaviour is correct even if suboptimal.
(I tested and it seems the case even if the stream is created with the `hipStreamNonBlocking` flag)

> Another way is to enqueue a host task into the stream which is executing `hipFree`. Then it will be asynchronous too.

I agree that this would be nicer, but maybe it's not worth the extra complexity :-/

I asked about this to the AMD people at LUMI and their answer is that 

> The async versions of memory allocation were only officially supported in ROCm 5.3. My opinion is that you shouldn't be doing any efforts around this for ROCm 5.2.3 and move

to a newer version of ROCm.

I'll update the PR accordingly.